# Titanic competition with TensorFlow Decision Forests

This notebook will take you through the steps needed to train a baseline Gradient Boosted Trees Model using TensorFlow Decision Forests and creating a submission on the Titanic competition. 

This notebook shows:

1. How to do some basic pre-processing. For example, the passenger names will be tokenized, and ticket names will be splitted in parts.
1. How to train a Gradient Boosted Trees (GBT) with default parameters
1. How to train a GBT with improved default parameters
1. How to tune the parameters of a GBTs
1. How to train and ensemble many GBTs

# Imports dependencies

In [12]:
import numpy as np
import pandas as pd
import os

import tensorflow as tf
import tensorflow_decision_forests as tfdf

print(f"Found TF-DF {tfdf.__version__}")

Found TF-DF 1.2.0


# Load dataset

In [13]:
train_df = pd.read_csv("/kaggle/input/titanic/train.csv")
serving_df = pd.read_csv("/kaggle/input/titanic/test.csv")

train_df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

# Prepare dataset

We will apply the following transformations on the dataset.

1. Tokenize the names. For example, "Braund, Mr. Owen Harris" will become ["Braund", "Mr.", "Owen", "Harris"].
2. Extract any prefix in the ticket. For example ticket "STON/O2. 3101282" will become "STON/O2." and 3101282.

In [14]:
def preprocess(df):
    df = df.copy()
    
    def normalize_name(x):
        return " ".join([v.strip(",()[].\"'") for v in x.split(" ")])
    
    def ticket_number(x):
        return x.split(" ")[-1]
        
    def ticket_item(x):
        items = x.split(" ")
        if len(items) == 1:
            return "NONE"
        return "_".join(items[0:-1])
    
    df["Name"] = df["Name"].apply(normalize_name)
    df["Ticket_number"] = df["Ticket"].apply(ticket_number)
    df["Ticket_item"] = df["Ticket"].apply(ticket_item)                     
    return df
    
preprocessed_train_df = preprocess(train_df)
preprocessed_serving_df = preprocess(serving_df)

preprocessed_train_df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                              Name     Sex   Age  SibSp  \
0                            Braund Mr Owen Harris    male  22.0      1   
1  Cumings Mrs John Bradley Florence Briggs Thayer  female  38.0      1   
2                             Heikkinen Miss Laina  female  26.0      0   
3         Futrelle Mrs Jacques Heath Lily May Peel  female  35.0      1   
4                           Allen Mr William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked Ticket_number Ticket_item  
0      0         A/5 21171   7.2500   NaN        S         21171         A/5  
1      0          PC 17599  71.2833   C85        C         17599          PC  
2      0  STON/O2. 3101282   7.9250   NaN        S       3101282    STON/O2.  
3      0            113803  53.1000  C123        S        113803        NONE  
4      0            373450   8.0500   NaN        S        373450        NONE

Let's keep the list of the input features of the model. Notably, we don't want to train our model on the "PassengerId" and "Ticket" features.

In [15]:
input_features = list(preprocessed_train_df.columns)
input_features.remove("Ticket")
input_features.remove("PassengerId")
input_features.remove("Survived")
#input_features.remove("Ticket_number")

print(f"Input features: {input_features}")

Input features: ['Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Cabin', 'Embarked', 'Ticket_number', 'Ticket_item']


# Convert Pandas dataset to TensorFlow Dataset

In [16]:
def tokenize_names(features, labels=None):
    """Divite the names into tokens. TF-DF can consume text tokens natively."""
    features["Name"] =  tf.strings.split(features["Name"])
    return features, labels

train_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_train_df,label="Survived").map(tokenize_names)
serving_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_serving_df).map(tokenize_names)

# Train model with default parameters

### Train model

First, we are training a GradientBoostedTreesModel model with the default parameters.

In [17]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    random_seed=1234,
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2024-01-05T13:13:34.753980236+00:00 kernel.cc:1214] Loading model from path /tmp/tmprvmmdk6l/model/ with prefix e63cb683520444d0
[INFO 2024-01-05T13:13:34.760707992+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-01-05T13:13:34.760765144+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.8260869383811951 Loss:0.8608942627906799


# Train model with improved default parameters

Now you'll use some specific parameters when creating the GBT model

In [18]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    
    #num_trees=2000,
    
    # Only for GBT.
    # A bit slower, but great to understand the model.
    # compute_permutation_variable_importance=True,
    
    # Change the default hyper-parameters
    # hyperparameter_template="benchmark_rank1@v1",
    
    #num_trees=1000,
    #tuner=tuner
    
    min_examples=1,
    categorical_algorithm="RANDOM",
    #max_depth=4,
    shrinkage=0.05,
    #num_candidate_attributes_ratio=0.2,
    split_axis="SPARSE_OBLIQUE",
    sparse_oblique_normalization="MIN_MAX",
    sparse_oblique_num_projections_exponent=2.0,
    num_trees=2000,
    #validation_ratio=0.0,
    random_seed=1234,
    
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2024-01-05T13:13:36.206283355+00:00 kernel.cc:1214] Loading model from path /tmp/tmp26y87yde/model/ with prefix 62ce131a9e5c4407
[INFO 2024-01-05T13:13:36.215539495+00:00 decision_forest.cc:661] Model loaded with 33 root(s), 1823 node(s), and 10 input feature(s).
[INFO 2024-01-05T13:13:36.215592247+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.760869562625885 Loss:1.0154211521148682


Let's look at the model and you can also notice the information about variable importance that the model figured out

In [19]:
model.summary()

Model: "gradient_boosted_trees_model_104"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
Total params: 1
Trainable params: 0
Non-trainable params: 1
_________________________________________________________________
Type: "GRADIENT_BOOSTED_TREES"
Task: CLASSIFICATION
Label: "__LABEL"

Input Features (11):
	Age
	Cabin
	Embarked
	Fare
	Name
	Parch
	Pclass
	Sex
	SibSp
	Ticket_item
	Ticket_number

No weights

Variable Importance: INV_MEAN_MIN_DEPTH:
    1.           "Sex"  0.576632 ################
    2.           "Age"  0.364297 #######
    3.          "Fare"  0.278839 ####
    4.          "Name"  0.208548 #
    5. "Ticket_number"  0.180792 
    6.        "Pclass"  0.176962 
    7.         "Parch"  0.176659 
    8.   "Ticket_item"  0.175540 
    9.      "Embarked"  0.172339 
   10.         "SibSp"  0.170442 

Variable Importance: NUM_AS_ROOT:
    1.  "Sex" 28.000000 ################
    2. "Name"  5.00000

# Make predictions

In [20]:
def prediction_to_kaggle_format(model, threshold=0.5):
    proba_survive = model.predict(serving_ds, verbose=0)[:,0]
    return pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (proba_survive >= threshold).astype(int)
    })

def make_submission(kaggle_predictions):
    path="/kaggle/working/submission.csv"
    kaggle_predictions.to_csv(path, index=False)
    print(f"Submission exported to {path}")
    
kaggle_predictions = prediction_to_kaggle_format(model)
make_submission(kaggle_predictions)
!head /kaggle/working/submission.csv

Submission exported to /kaggle/working/submission.csv
PassengerId,Survived
892,0
893,0
894,0
895,0
896,0
897,0
898,0
899,0
900,1


# Training a model with hyperparameter tunning

Hyper-parameter tuning is enabled by specifying the tuner constructor argument of the model. The tuner object contains all the configuration of the tuner (search space, optimizer, trial and objective).


In [21]:
tuner = tfdf.tuner.RandomSearch(num_trials=1000)
tuner.choice("min_examples", [2, 5, 7, 10])
tuner.choice("categorical_algorithm", ["CART", "RANDOM"])

local_search_space = tuner.choice("growing_strategy", ["LOCAL"])
local_search_space.choice("max_depth", [3, 4, 5, 6, 8])

global_search_space = tuner.choice("growing_strategy", ["BEST_FIRST_GLOBAL"], merge=True)
global_search_space.choice("max_num_nodes", [16, 32, 64, 128, 256])

#tuner.choice("use_hessian_gain", [True, False])
tuner.choice("shrinkage", [0.02, 0.05, 0.10, 0.15])
tuner.choice("num_candidate_attributes_ratio", [0.2, 0.5, 0.9, 1.0])


tuner.choice("split_axis", ["AXIS_ALIGNED"])
oblique_space = tuner.choice("split_axis", ["SPARSE_OBLIQUE"], merge=True)
oblique_space.choice("sparse_oblique_normalization",
                     ["NONE", "STANDARD_DEVIATION", "MIN_MAX"])
oblique_space.choice("sparse_oblique_weights", ["BINARY", "CONTINUOUS"])
oblique_space.choice("sparse_oblique_num_projections_exponent", [1.0, 1.5])

# Tune the model. Notice the `tuner=tuner`.
tuned_model = tfdf.keras.GradientBoostedTreesModel(tuner=tuner)
tuned_model.fit(train_ds, verbose=0)

tuned_self_evaluation = tuned_model.make_inspector().evaluation()
print(f"Accuracy: {tuned_self_evaluation.accuracy} Loss:{tuned_self_evaluation.loss}")

Use /tmp/tmpr0qn45ow as temporary training directory


[INFO 2024-01-05T13:15:45.797374038+00:00 kernel.cc:1214] Loading model from path /tmp/tmpr0qn45ow/model/ with prefix bc90212c088d4738
[INFO 2024-01-05T13:15:45.815946404+00:00 decision_forest.cc:661] Model loaded with 19 root(s), 589 node(s), and 12 input feature(s).
[INFO 2024-01-05T13:15:45.816010095+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesGeneric" built
[INFO 2024-01-05T13:15:45.816049788+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.9178082346916199 Loss:0.6503586769104004


In the last line in the cell above, you can see the accuracy is higher than previously with default parameters and parameters set by hand.

This is the main idea behing hyperparameter tuning.

For more information you can follow this tutorial: [Automated hyper-parameter tuning](https://www.tensorflow.org/decision_forests/tutorials/automatic_tuning_colab)

# Making an ensemble

Here you'll create 100 models with different seeds and combine their results

This approach removes a little bit the random aspects related to creating ML models

In the GBT creation is used the `honest` parameter. It will use different training examples to infer the structure and the leaf values. This regularization technique trades examples for bias estimates.

In [22]:
predictions = None
num_predictions = 0

for i in range(100):
    print(f"i:{i}")
    # Possible models: GradientBoostedTreesModel or RandomForestModel
    model = tfdf.keras.GradientBoostedTreesModel(
        verbose=0, # Very few logs
        features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
        exclude_non_specified_features=True, # Only use the features in "features"

        #min_examples=1,
        #categorical_algorithm="RANDOM",
        ##max_depth=4,
        #shrinkage=0.05,
        ##num_candidate_attributes_ratio=0.2,
        #split_axis="SPARSE_OBLIQUE",
        #sparse_oblique_normalization="MIN_MAX",
        #sparse_oblique_num_projections_exponent=2.0,
        #num_trees=2000,
        ##validation_ratio=0.0,
        random_seed=i,
        honest=True,
    )
    model.fit(train_ds)
    
    sub_predictions = model.predict(serving_ds, verbose=0)[:,0]
    if predictions is None:
        predictions = sub_predictions
    else:
        predictions += sub_predictions
    num_predictions += 1

predictions/=num_predictions

kaggle_predictions = pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (predictions >= 0.5).astype(int)
    })

make_submission(kaggle_predictions)

i:0


[INFO 2024-01-05T13:15:46.919747188+00:00 kernel.cc:1214] Loading model from path /tmp/tmpw3tb07ix/model/ with prefix cc210108d7fa4702
[INFO 2024-01-05T13:15:46.924790689+00:00 kernel.cc:1046] Use fast generic engine


i:1


[INFO 2024-01-05T13:15:48.488755864+00:00 kernel.cc:1214] Loading model from path /tmp/tmpd25viz3g/model/ with prefix ad07ac0fd8ca4c99
[INFO 2024-01-05T13:15:48.509875129+00:00 kernel.cc:1046] Use fast generic engine


i:2


[INFO 2024-01-05T13:15:49.675415502+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2elq_2p8/model/ with prefix 947ad0cbaf3d41cd
[INFO 2024-01-05T13:15:49.680362595+00:00 kernel.cc:1046] Use fast generic engine


i:3


[INFO 2024-01-05T13:15:51.439574311+00:00 kernel.cc:1214] Loading model from path /tmp/tmp012rg3cn/model/ with prefix 584a77c63f7248be
[INFO 2024-01-05T13:15:51.468127088+00:00 kernel.cc:1046] Use fast generic engine


i:4


[INFO 2024-01-05T13:15:52.660271228+00:00 kernel.cc:1214] Loading model from path /tmp/tmpga8e4ilz/model/ with prefix 15b4e6d6d2614e2f
[INFO 2024-01-05T13:15:52.66653798+00:00 kernel.cc:1046] Use fast generic engine


i:5


[INFO 2024-01-05T13:15:53.788307074+00:00 kernel.cc:1214] Loading model from path /tmp/tmprftwi0sx/model/ with prefix f188f2a8b76e468a
[INFO 2024-01-05T13:15:53.791797405+00:00 kernel.cc:1046] Use fast generic engine


i:6


[INFO 2024-01-05T13:15:55.052086638+00:00 kernel.cc:1214] Loading model from path /tmp/tmpr09misby/model/ with prefix 2896e6f6ae424016
[INFO 2024-01-05T13:15:55.060484709+00:00 kernel.cc:1046] Use fast generic engine


i:7


[INFO 2024-01-05T13:15:56.680428741+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxmupydk7/model/ with prefix 804bb51850884447
[INFO 2024-01-05T13:15:56.703654883+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-01-05T13:15:56.703705279+00:00 kernel.cc:1046] Use fast generic engine


i:8


[INFO 2024-01-05T13:15:58.196123514+00:00 kernel.cc:1214] Loading model from path /tmp/tmpd8eynniw/model/ with prefix 108cfb930d4f42fb
[INFO 2024-01-05T13:15:58.206689322+00:00 kernel.cc:1046] Use fast generic engine


i:9


[INFO 2024-01-05T13:15:59.677283903+00:00 kernel.cc:1214] Loading model from path /tmp/tmpihn1qr9q/model/ with prefix 86caa18d6b054bf7
[INFO 2024-01-05T13:15:59.69321943+00:00 kernel.cc:1046] Use fast generic engine


i:10


[INFO 2024-01-05T13:16:01.90628179+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxeyz5pj0/model/ with prefix 7631073ca60b4b62
[INFO 2024-01-05T13:16:01.914457578+00:00 kernel.cc:1046] Use fast generic engine


i:11


[INFO 2024-01-05T13:16:04.06346152+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmkvs1edq/model/ with prefix 2835005eb628448b
[INFO 2024-01-05T13:16:04.082080288+00:00 kernel.cc:1046] Use fast generic engine


i:12


[INFO 2024-01-05T13:16:05.695383284+00:00 kernel.cc:1214] Loading model from path /tmp/tmptmsfcn6b/model/ with prefix c82c9f18bb794d1b
[INFO 2024-01-05T13:16:05.702439757+00:00 kernel.cc:1046] Use fast generic engine


i:13


[INFO 2024-01-05T13:16:07.698248103+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7p044835/model/ with prefix 33000533fafa4959
[INFO 2024-01-05T13:16:07.710836632+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-01-05T13:16:07.710882944+00:00 kernel.cc:1046] Use fast generic engine


i:14


[INFO 2024-01-05T13:16:09.425536828+00:00 kernel.cc:1214] Loading model from path /tmp/tmp84mlx0f3/model/ with prefix 1db5aaaac55f4a15
[INFO 2024-01-05T13:16:09.43326081+00:00 kernel.cc:1046] Use fast generic engine


i:15


[INFO 2024-01-05T13:16:10.715876486+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_oizk0s2/model/ with prefix d3bb8ed3f2044185
[INFO 2024-01-05T13:16:10.724449466+00:00 kernel.cc:1046] Use fast generic engine


i:16


[INFO 2024-01-05T13:16:12.124017701+00:00 kernel.cc:1214] Loading model from path /tmp/tmpf7aemuj4/model/ with prefix e40173c20a8d461a
[INFO 2024-01-05T13:16:12.138330098+00:00 kernel.cc:1046] Use fast generic engine


i:17


[INFO 2024-01-05T13:16:13.56420159+00:00 kernel.cc:1214] Loading model from path /tmp/tmp88i03jaw/model/ with prefix 6a994f0b74e94165
[INFO 2024-01-05T13:16:13.578877399+00:00 kernel.cc:1046] Use fast generic engine


i:18


[INFO 2024-01-05T13:16:14.985266628+00:00 kernel.cc:1214] Loading model from path /tmp/tmpai9e91ca/model/ with prefix ba0e667d0e664493
[INFO 2024-01-05T13:16:14.998047892+00:00 kernel.cc:1046] Use fast generic engine


i:19


[INFO 2024-01-05T13:16:16.568781694+00:00 kernel.cc:1214] Loading model from path /tmp/tmpyjk4vqls/model/ with prefix 262a7fa874ff4682
[INFO 2024-01-05T13:16:16.588778074+00:00 kernel.cc:1046] Use fast generic engine


i:20


[INFO 2024-01-05T13:16:18.309633286+00:00 kernel.cc:1214] Loading model from path /tmp/tmppc1ivqks/model/ with prefix 4e432734af4b4cc7
[INFO 2024-01-05T13:16:18.326309759+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-01-05T13:16:18.326354481+00:00 kernel.cc:1046] Use fast generic engine


i:21


[INFO 2024-01-05T13:16:19.850899814+00:00 kernel.cc:1214] Loading model from path /tmp/tmp483t4ahg/model/ with prefix 6ca58f05b8bc49b8
[INFO 2024-01-05T13:16:19.856581346+00:00 kernel.cc:1046] Use fast generic engine


i:22


[INFO 2024-01-05T13:16:21.057298358+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5idaajyn/model/ with prefix 59b47be27b834b23
[INFO 2024-01-05T13:16:21.063629692+00:00 kernel.cc:1046] Use fast generic engine


i:23


[INFO 2024-01-05T13:16:22.979238384+00:00 kernel.cc:1214] Loading model from path /tmp/tmpibc88ri6/model/ with prefix e8a90459612e42ae
[INFO 2024-01-05T13:16:22.989220798+00:00 kernel.cc:1046] Use fast generic engine


i:24


[INFO 2024-01-05T13:16:24.190165981+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmcjp9poa/model/ with prefix 2f679be51be84d66
[INFO 2024-01-05T13:16:24.196724655+00:00 kernel.cc:1046] Use fast generic engine


i:25


[INFO 2024-01-05T13:16:25.578189648+00:00 kernel.cc:1214] Loading model from path /tmp/tmppj8p9vog/model/ with prefix 06f408b05b684b0b
[INFO 2024-01-05T13:16:25.590774076+00:00 kernel.cc:1046] Use fast generic engine


i:26


[INFO 2024-01-05T13:16:26.926864844+00:00 kernel.cc:1214] Loading model from path /tmp/tmp406phf6c/model/ with prefix a24b21e72a714025
[INFO 2024-01-05T13:16:26.937641868+00:00 kernel.cc:1046] Use fast generic engine


i:27


[INFO 2024-01-05T13:16:28.651644153+00:00 kernel.cc:1214] Loading model from path /tmp/tmprvlwptjx/model/ with prefix fc32a418574d40bf
[INFO 2024-01-05T13:16:28.658789591+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-01-05T13:16:28.658848891+00:00 kernel.cc:1046] Use fast generic engine


i:28


[INFO 2024-01-05T13:16:29.891487332+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2w33gfx6/model/ with prefix e333beeff79e4820
[INFO 2024-01-05T13:16:29.89735229+00:00 kernel.cc:1046] Use fast generic engine


i:29


[INFO 2024-01-05T13:16:31.323323316+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfjz_75cs/model/ with prefix 83211634fbd54f76
[INFO 2024-01-05T13:16:31.337326424+00:00 kernel.cc:1046] Use fast generic engine


i:30


[INFO 2024-01-05T13:16:33.525129401+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3kyfqn18/model/ with prefix 00328b4aa9b8414b
[INFO 2024-01-05T13:16:33.555700114+00:00 kernel.cc:1046] Use fast generic engine


i:31


[INFO 2024-01-05T13:16:34.926241751+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0aoy_8r8/model/ with prefix d4479c2c990e423b
[INFO 2024-01-05T13:16:34.936443156+00:00 kernel.cc:1046] Use fast generic engine


i:32


[INFO 2024-01-05T13:16:36.147895863+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5ygirvzd/model/ with prefix 68d097d9fad840cf
[INFO 2024-01-05T13:16:36.15417593+00:00 kernel.cc:1046] Use fast generic engine


i:33


[INFO 2024-01-05T13:16:37.585542323+00:00 kernel.cc:1214] Loading model from path /tmp/tmphe8ihi2_/model/ with prefix ef2e07e4d3054a18
[INFO 2024-01-05T13:16:37.599765496+00:00 kernel.cc:1046] Use fast generic engine


i:34


[INFO 2024-01-05T13:16:38.897912171+00:00 kernel.cc:1214] Loading model from path /tmp/tmpa8dztan4/model/ with prefix e7c4e182d0d44599
[INFO 2024-01-05T13:16:38.907587674+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-01-05T13:16:38.907649177+00:00 kernel.cc:1046] Use fast generic engine


i:35


[INFO 2024-01-05T13:16:40.277991932+00:00 kernel.cc:1214] Loading model from path /tmp/tmpts8c3awg/model/ with prefix a4fcc44e5c6444b5
[INFO 2024-01-05T13:16:40.287035514+00:00 kernel.cc:1046] Use fast generic engine


i:36


[INFO 2024-01-05T13:16:41.748660347+00:00 kernel.cc:1214] Loading model from path /tmp/tmpw8b9f9k5/model/ with prefix 4aa21a537629433d
[INFO 2024-01-05T13:16:41.764393546+00:00 kernel.cc:1046] Use fast generic engine


i:37


[INFO 2024-01-05T13:16:43.434862468+00:00 kernel.cc:1214] Loading model from path /tmp/tmp87fzdxfp/model/ with prefix 637b201a65c142e5
[INFO 2024-01-05T13:16:43.444389225+00:00 kernel.cc:1046] Use fast generic engine


i:38


[INFO 2024-01-05T13:16:44.936831089+00:00 kernel.cc:1214] Loading model from path /tmp/tmptqmqkq0p/model/ with prefix a08a3524b71b4a3f
[INFO 2024-01-05T13:16:44.952118345+00:00 kernel.cc:1046] Use fast generic engine


i:39


[INFO 2024-01-05T13:16:46.390448934+00:00 kernel.cc:1214] Loading model from path /tmp/tmpiglsnknj/model/ with prefix f90b10a0873e49fd
[INFO 2024-01-05T13:16:46.404753531+00:00 kernel.cc:1046] Use fast generic engine


i:40


[INFO 2024-01-05T13:16:47.597426561+00:00 kernel.cc:1214] Loading model from path /tmp/tmp71e1v0zv/model/ with prefix 708e9805a913405e
[INFO 2024-01-05T13:16:47.602418302+00:00 kernel.cc:1046] Use fast generic engine


i:41


[INFO 2024-01-05T13:16:49.103848273+00:00 kernel.cc:1214] Loading model from path /tmp/tmpaixjy2uy/model/ with prefix 1196c9565b3b407f
[INFO 2024-01-05T13:16:49.120603615+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-01-05T13:16:49.120649199+00:00 kernel.cc:1046] Use fast generic engine


i:42


[INFO 2024-01-05T13:16:50.474169025+00:00 kernel.cc:1214] Loading model from path /tmp/tmpv7e5bx9g/model/ with prefix 10832fa01c784d0a
[INFO 2024-01-05T13:16:50.484176324+00:00 kernel.cc:1046] Use fast generic engine


i:43


[INFO 2024-01-05T13:16:52.047125151+00:00 kernel.cc:1214] Loading model from path /tmp/tmprm6_nvkm/model/ with prefix cece91379bff4149
[INFO 2024-01-05T13:16:52.065849948+00:00 kernel.cc:1046] Use fast generic engine


i:44


[INFO 2024-01-05T13:16:54.35750398+00:00 kernel.cc:1214] Loading model from path /tmp/tmpy6mg3i52/model/ with prefix ac7d2f77c7fb4e5c
[INFO 2024-01-05T13:16:54.369646564+00:00 kernel.cc:1046] Use fast generic engine


i:45


[INFO 2024-01-05T13:16:56.10375769+00:00 kernel.cc:1214] Loading model from path /tmp/tmpl9b1e_h2/model/ with prefix 2e02e232ceb740bd
[INFO 2024-01-05T13:16:56.108131424+00:00 kernel.cc:1046] Use fast generic engine


i:46


[INFO 2024-01-05T13:16:57.655082188+00:00 kernel.cc:1214] Loading model from path /tmp/tmpluuw02z4/model/ with prefix b3b970004bdd4e43
[INFO 2024-01-05T13:16:57.672456595+00:00 kernel.cc:1046] Use fast generic engine


i:47


[INFO 2024-01-05T13:16:59.489376597+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5x4_wa92/model/ with prefix 56a398462911403b
[INFO 2024-01-05T13:16:59.506476066+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-01-05T13:16:59.506526798+00:00 kernel.cc:1046] Use fast generic engine


i:48


[INFO 2024-01-05T13:17:00.724444101+00:00 kernel.cc:1214] Loading model from path /tmp/tmpx03r7rd4/model/ with prefix ac1e65aa05514556
[INFO 2024-01-05T13:17:00.729840495+00:00 kernel.cc:1046] Use fast generic engine


i:49


[INFO 2024-01-05T13:17:02.021724465+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgx5_86dz/model/ with prefix 6fcefcf3fdcc4bea
[INFO 2024-01-05T13:17:02.029783041+00:00 kernel.cc:1046] Use fast generic engine


i:50


[INFO 2024-01-05T13:17:03.457921594+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfzozosno/model/ with prefix df6f6ca3affe480e
[INFO 2024-01-05T13:17:03.471629788+00:00 kernel.cc:1046] Use fast generic engine


i:51


[INFO 2024-01-05T13:17:05.00728652+00:00 kernel.cc:1214] Loading model from path /tmp/tmpa3c49731/model/ with prefix f9a0b46a5d644eee
[INFO 2024-01-05T13:17:05.024792295+00:00 kernel.cc:1046] Use fast generic engine


i:52


[INFO 2024-01-05T13:17:06.307254572+00:00 kernel.cc:1214] Loading model from path /tmp/tmpo3izq959/model/ with prefix 602830049dde4f46
[INFO 2024-01-05T13:17:06.315970424+00:00 kernel.cc:1046] Use fast generic engine


i:53


[INFO 2024-01-05T13:17:07.578975711+00:00 kernel.cc:1214] Loading model from path /tmp/tmphda6izir/model/ with prefix 3d50cef315de4679
[INFO 2024-01-05T13:17:07.587083579+00:00 kernel.cc:1046] Use fast generic engine


i:54


[INFO 2024-01-05T13:17:08.749199142+00:00 kernel.cc:1214] Loading model from path /tmp/tmpz9owrtyv/model/ with prefix 3638f0679ee04903
[INFO 2024-01-05T13:17:08.752725325+00:00 kernel.cc:1046] Use fast generic engine


i:55


[INFO 2024-01-05T13:17:10.308592105+00:00 kernel.cc:1214] Loading model from path /tmp/tmpktd_35jg/model/ with prefix 31254fcb2d5c4ad4
[INFO 2024-01-05T13:17:10.324521805+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-01-05T13:17:10.324573711+00:00 kernel.cc:1046] Use fast generic engine


i:56


[INFO 2024-01-05T13:17:12.088512479+00:00 kernel.cc:1214] Loading model from path /tmp/tmptpc1db9x/model/ with prefix 617c303485f2416b
[INFO 2024-01-05T13:17:12.100501714+00:00 kernel.cc:1046] Use fast generic engine


i:57


[INFO 2024-01-05T13:17:13.316128632+00:00 kernel.cc:1214] Loading model from path /tmp/tmp45b1eniw/model/ with prefix f10d78f2733547f4
[INFO 2024-01-05T13:17:13.320796561+00:00 kernel.cc:1046] Use fast generic engine


i:58


[INFO 2024-01-05T13:17:14.578618252+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgmepmjek/model/ with prefix 5824742d23c541f1
[INFO 2024-01-05T13:17:14.58791876+00:00 kernel.cc:1046] Use fast generic engine


i:59


[INFO 2024-01-05T13:17:15.951792889+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmxrytrjl/model/ with prefix 99f4908076d74a41
[INFO 2024-01-05T13:17:15.962046098+00:00 kernel.cc:1046] Use fast generic engine


i:60


[INFO 2024-01-05T13:17:17.323405292+00:00 kernel.cc:1214] Loading model from path /tmp/tmpub6j298v/model/ with prefix 23ba893e6ac4481c
[INFO 2024-01-05T13:17:17.334267676+00:00 kernel.cc:1046] Use fast generic engine


i:61


[INFO 2024-01-05T13:17:18.542899047+00:00 kernel.cc:1214] Loading model from path /tmp/tmpok7j2jah/model/ with prefix d455a42db903470a
[INFO 2024-01-05T13:17:18.548385167+00:00 kernel.cc:1046] Use fast generic engine


i:62


[INFO 2024-01-05T13:17:20.353129794+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7gqah00w/model/ with prefix aa1839ebcf584de3
[INFO 2024-01-05T13:17:20.381606618+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-01-05T13:17:20.381656049+00:00 kernel.cc:1046] Use fast generic engine


i:63


[INFO 2024-01-05T13:17:23.012079704+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2sxz7mr3/model/ with prefix bf217ed973d44e96
[INFO 2024-01-05T13:17:23.022340949+00:00 kernel.cc:1046] Use fast generic engine


i:64


[INFO 2024-01-05T13:17:24.363103799+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3e7fypag/model/ with prefix 861fff833a394603
[INFO 2024-01-05T13:17:24.371810882+00:00 kernel.cc:1046] Use fast generic engine


i:65


[INFO 2024-01-05T13:17:25.679663263+00:00 kernel.cc:1214] Loading model from path /tmp/tmphudjcvb1/model/ with prefix d9fc193e251548b3
[INFO 2024-01-05T13:17:25.685121125+00:00 kernel.cc:1046] Use fast generic engine


i:66


[INFO 2024-01-05T13:17:27.024002304+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsz5skbv7/model/ with prefix 098d948643c54407
[INFO 2024-01-05T13:17:27.031338498+00:00 kernel.cc:1046] Use fast generic engine


i:67


[INFO 2024-01-05T13:17:28.683864178+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwnbpciya/model/ with prefix a65296e553074dd0
[INFO 2024-01-05T13:17:28.704424708+00:00 kernel.cc:1046] Use fast generic engine


i:68


[INFO 2024-01-05T13:17:30.199847703+00:00 kernel.cc:1214] Loading model from path /tmp/tmpj980g88w/model/ with prefix 05d767f3a9504846
[INFO 2024-01-05T13:17:30.212610209+00:00 kernel.cc:1046] Use fast generic engine


i:69


[INFO 2024-01-05T13:17:31.565474333+00:00 kernel.cc:1214] Loading model from path /tmp/tmpu1pm71xt/model/ with prefix 88719f1245994a31
[INFO 2024-01-05T13:17:31.572969783+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-01-05T13:17:31.573032558+00:00 kernel.cc:1046] Use fast generic engine


i:70


[INFO 2024-01-05T13:17:33.492308817+00:00 kernel.cc:1214] Loading model from path /tmp/tmpyjmh7epw/model/ with prefix 2178cdca26ce4289
[INFO 2024-01-05T13:17:33.502146794+00:00 kernel.cc:1046] Use fast generic engine


i:71


[INFO 2024-01-05T13:17:34.8433169+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_horsy78/model/ with prefix fd3749903342453d
[INFO 2024-01-05T13:17:34.851693857+00:00 kernel.cc:1046] Use fast generic engine


i:72


[INFO 2024-01-05T13:17:36.45702584+00:00 kernel.cc:1214] Loading model from path /tmp/tmp45f9ytql/model/ with prefix f8e8517e17254407
[INFO 2024-01-05T13:17:36.47409007+00:00 kernel.cc:1046] Use fast generic engine


i:73


[INFO 2024-01-05T13:17:37.767569672+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1ii1kiq1/model/ with prefix f8f569eaa8f7492a
[INFO 2024-01-05T13:17:37.775335559+00:00 kernel.cc:1046] Use fast generic engine


i:74


[INFO 2024-01-05T13:17:39.789981988+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgjjk3dr7/model/ with prefix d7128074f9b14363
[INFO 2024-01-05T13:17:39.802811536+00:00 kernel.cc:1046] Use fast generic engine


i:75


[INFO 2024-01-05T13:17:41.133067976+00:00 kernel.cc:1214] Loading model from path /tmp/tmps4wvddq8/model/ with prefix 15fbcd376dfa4fe6
[INFO 2024-01-05T13:17:41.141948763+00:00 kernel.cc:1046] Use fast generic engine


i:76


[INFO 2024-01-05T13:17:42.348470118+00:00 kernel.cc:1214] Loading model from path /tmp/tmpajyr5ja0/model/ with prefix f9ffe9a570b6457e
[INFO 2024-01-05T13:17:42.35271103+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-01-05T13:17:42.352749949+00:00 kernel.cc:1046] Use fast generic engine


i:77


[INFO 2024-01-05T13:17:43.64345224+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfzyvgbt_/model/ with prefix 68686aa8c70b4142
[INFO 2024-01-05T13:17:43.649245461+00:00 kernel.cc:1046] Use fast generic engine


i:78


[INFO 2024-01-05T13:17:44.949916037+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnnn4wjzg/model/ with prefix 6e246a60f304499e
[INFO 2024-01-05T13:17:44.957428324+00:00 kernel.cc:1046] Use fast generic engine


i:79


[INFO 2024-01-05T13:17:47.239190498+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2gzjlthn/model/ with prefix dbdfb83a6bb34a68
[INFO 2024-01-05T13:17:47.249454815+00:00 kernel.cc:1046] Use fast generic engine


i:80


[INFO 2024-01-05T13:17:48.744372829+00:00 kernel.cc:1214] Loading model from path /tmp/tmpiicxf_fd/model/ with prefix 3fe1432385fb4fd5
[INFO 2024-01-05T13:17:48.756939305+00:00 kernel.cc:1046] Use fast generic engine


i:81


[INFO 2024-01-05T13:17:50.249405479+00:00 kernel.cc:1214] Loading model from path /tmp/tmpse6au3o5/model/ with prefix 393a1593f5014d6e
[INFO 2024-01-05T13:17:50.264019819+00:00 kernel.cc:1046] Use fast generic engine


i:82


[INFO 2024-01-05T13:17:51.666635871+00:00 kernel.cc:1214] Loading model from path /tmp/tmpr22rpx_v/model/ with prefix 44147acef4b94472
[INFO 2024-01-05T13:17:51.678029353+00:00 kernel.cc:1046] Use fast generic engine


i:83


[INFO 2024-01-05T13:17:53.542795686+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3g31rq5u/model/ with prefix a6f60ab7c4ef4754
[INFO 2024-01-05T13:17:53.552889798+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-01-05T13:17:53.552949857+00:00 kernel.cc:1046] Use fast generic engine


i:84


[INFO 2024-01-05T13:17:55.135088945+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkf160lyj/model/ with prefix 750e3b370aea4a0a
[INFO 2024-01-05T13:17:55.15250295+00:00 kernel.cc:1046] Use fast generic engine


i:85


[INFO 2024-01-05T13:17:56.430338288+00:00 kernel.cc:1214] Loading model from path /tmp/tmpq6qcdgbw/model/ with prefix d2976881fe1c48ed
[INFO 2024-01-05T13:17:56.43762117+00:00 kernel.cc:1046] Use fast generic engine


i:86


[INFO 2024-01-05T13:17:57.964502477+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7cew2o2c/model/ with prefix c002ecf89d734e2a
[INFO 2024-01-05T13:17:57.981908645+00:00 kernel.cc:1046] Use fast generic engine


i:87


[INFO 2024-01-05T13:17:59.576553719+00:00 kernel.cc:1214] Loading model from path /tmp/tmpksp_ldno/model/ with prefix a969a136fdd1406b
[INFO 2024-01-05T13:17:59.596203323+00:00 kernel.cc:1046] Use fast generic engine


i:88


[INFO 2024-01-05T13:18:01.033243609+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvptilwh3/model/ with prefix f5914f2400184cd0
[INFO 2024-01-05T13:18:01.046297827+00:00 kernel.cc:1046] Use fast generic engine


i:89


[INFO 2024-01-05T13:18:02.321265735+00:00 kernel.cc:1214] Loading model from path /tmp/tmpl_j9dho8/model/ with prefix 3b43bc2026c64318
[INFO 2024-01-05T13:18:02.326594057+00:00 kernel.cc:1046] Use fast generic engine


i:90


[INFO 2024-01-05T13:18:03.717343148+00:00 kernel.cc:1214] Loading model from path /tmp/tmpu_akcnce/model/ with prefix 3578c8bc7a904bd8
[INFO 2024-01-05T13:18:03.728713196+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-01-05T13:18:03.728771205+00:00 kernel.cc:1046] Use fast generic engine


i:91


[INFO 2024-01-05T13:18:05.504077921+00:00 kernel.cc:1214] Loading model from path /tmp/tmpskykdrr4/model/ with prefix 719752f9f44f4b12
[INFO 2024-01-05T13:18:05.511593282+00:00 kernel.cc:1046] Use fast generic engine


i:92


[INFO 2024-01-05T13:18:07.724319391+00:00 kernel.cc:1214] Loading model from path /tmp/tmppbzwlvnv/model/ with prefix 01fd30d3ee7a40a0
[INFO 2024-01-05T13:18:07.744231366+00:00 kernel.cc:1046] Use fast generic engine


i:93


[INFO 2024-01-05T13:18:09.40302515+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfs_jfvfd/model/ with prefix 988455e3dbd74dd7
[INFO 2024-01-05T13:18:09.413981751+00:00 kernel.cc:1046] Use fast generic engine


i:94


[INFO 2024-01-05T13:18:10.728823344+00:00 kernel.cc:1214] Loading model from path /tmp/tmprhvzh0i5/model/ with prefix 7efd3e699ccb49e4
[INFO 2024-01-05T13:18:10.736014606+00:00 kernel.cc:1046] Use fast generic engine


i:95


[INFO 2024-01-05T13:18:12.076305104+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsatoupo0/model/ with prefix dc363de010044a86
[INFO 2024-01-05T13:18:12.085505874+00:00 kernel.cc:1046] Use fast generic engine


i:96


[INFO 2024-01-05T13:18:13.460690087+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqq31bp3i/model/ with prefix 58870f0072e143df
[INFO 2024-01-05T13:18:13.471959892+00:00 kernel.cc:1046] Use fast generic engine


i:97


[INFO 2024-01-05T13:18:14.809588009+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcgs8j_5k/model/ with prefix 7bfdd23065104704
[INFO 2024-01-05T13:18:14.81550561+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-01-05T13:18:14.815563266+00:00 kernel.cc:1046] Use fast generic engine


i:98


[INFO 2024-01-05T13:18:16.679995588+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvt8wqwwg/model/ with prefix c7767b5da7464d3f
[INFO 2024-01-05T13:18:16.688495963+00:00 kernel.cc:1046] Use fast generic engine


i:99


[INFO 2024-01-05T13:18:18.188118756+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfi48ntsb/model/ with prefix 42f3f2a67a8a4e00
[INFO 2024-01-05T13:18:18.203447321+00:00 kernel.cc:1046] Use fast generic engine


Submission exported to /kaggle/working/submission.csv


# What is next

If you want to learn more about TensorFlow Decision Forests and its advanced features, you can follow the official documentation [here](https://www.tensorflow.org/decision_forests) 